Ieee802154 switch to new net pkt API

tbursztyka opened the pull request:
Instead of using net_pkt_write in fragmentation, appending the frag to the packet is way more efficient (no allocation, de-allocation).

Commit:
drivers/ieee802154: Switch CC1200 to new net_ptk allocator

Now buffer is allocated at the same time.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ieee802154: Switch CC2520 to new net_ptk allocator

Now buffer is allocated at the same time.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ieee802154: Switch MCR20A to new net_ptk allocator

Now buffer is allocated at the same time.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ieee802154: Switch NRF5 to new net_ptk allocator

Now buffer is allocated at the same time.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
drivers/ieee802154: Switch KW41Z to new net_ptk allocator

Now buffer is allocated at the same time.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Switch L2 to new net_pkt allocator

Now buffer is allocated at the same time.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

Commit:
net/ieee802154: Switch fragmentation support to new net_pkt API

- differentiating actual "fragment" to buffer element from a net_pkt
- instead of copying data (and thus allocating buffer) let's just take
the fragment buffer and put it into result packet.
- fixing compilation issue in relevant test

It's more efficient that way, as we use already allocated fragment
buffer instead of reallocating/deallocating after each fragment
reception.

A possible optimization would be to calculate the actual size + header
difference if only the actual size is close to the target size.
It would avoid to get a dry run of the header decompression for each
fragment received. The difficulty being in finding the sweet spot when
it is relevant to calculate the header difference.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

rlubos reviewed (changes requested):
This PR does not build for nRF:

```
cmake -GNinja -DBOARD=nrf52840_pca10056 -DOVERLAY_CONFIG=overlay-802154.conf ..
```

```
/home/rolu/repos/zephyrproject/zephyr/drivers/ieee802154/ieee802154_nrf5.c:103:26: warning: passing argument 2 of ‘net_pkt_write’ from incompatible pointer type [-Wincompatible-pointer-types]
   if (net_pkt_write(pkt, rx_frame->psdu + 1, pkt_len)) {
                          ^~~~~~~~
In file included from /home/rolu/repos/zephyrproject/zephyr/drivers/ieee802154/ieee802154_nrf5.c:24:0:
../../../../../include/net/net_pkt.h:1695:17: note: expected ‘struct net_buf *’ but argument is of type ‘u8_t * {aka unsigned char *}’
 struct net_buf *net_pkt_write(struct net_pkt *pkt, struct net_buf *frag,
                 ^~~~~~~~~~~~~
/home/rolu/repos/zephyrproject/zephyr/drivers/ieee802154/ieee802154_nrf5.c:103:7: error: too few arguments to function ‘net_pkt_write’
   if (net_pkt_write(pkt, rx_frame->psdu + 1, pkt_len)) {
       ^~~~~~~~~~~~~
In file included from /home/rolu/repos/zephyrproject/zephyr/drivers/ieee802154/ieee802154_nrf5.c:24:0:
../../../../../include/net/net_pkt.h:1695:17: note: declared here
 struct net_buf *net_pkt_write(struct net_pkt *pkt, struct net_buf *frag,
```

Not sure why didn't CI did not catch this - this configuration was recently added to smoke tests.

tbursztyka commented:
indeed

rlubos reviewed (commented):
Thank you, nRF driver builds fine now. The communication seems to be working fine, I'm observing some errors from 802.15.4 L2 quite often though: 
```
[00:00:57.824,890] <err> net_ieee802154_fragment: Copying frag failed
[00:00:57.829,559] <err> net_ieee802154_fragment: Copying frag failed
[00:00:57.834,197] <err> net_ieee802154_fragment: Copying frag failed
[00:00:57.838,806] <err> net_ieee802154_fragment: Copying frag failed
[00:00:57.843,414] <err> net_ieee802154_fragment: Copying frag failed
```
I've quadrupled the default number of net packets/buffers  just to be sure it's not an allocation problem. Is this something to be concerned about?

tbursztyka commented:
@rlubos thing is now it does not copy anything (error message should be updated), so it can't be a memory allocation issue. But there is something fishy with comparing offset net_pkt_gen_len() now (see ieee802154_fragment.c:append_frag()

rlubos commented on subsys/net/l2/ieee802154/ieee802154_fragment.c:
```diff
@@ -334,34 +345,23 @@ static inline struct frag_cache *get_reass_cache(u16_t size, u16_t tag)
-			pkt->frags = NULL;
+	if (offset != net_pkt_get_len(pkt)) {
+		/* Given offset is wrong*/
+		NET_DBG("Offset %u vs current length %zu",
```
`NET_ERR` perhaps?

tbursztyka replied:
Yep

rveerama1 commented on subsys/net/l2/ieee802154/ieee802154_fragment.c:
```diff
@@ -334,34 +345,23 @@ static inline struct frag_cache *get_reass_cache(u16_t size, u16_t tag)
 	return NULL;
 }
 
-/* Helper function to write fragment data to Rx packet based on offset. */
```
Seems some confusion here. While receiving 15.4 fragments can come in any order. We should be able to assemble them based on their offset. Its not always appending the buffer. Original code handles that or I am missing something in these changes.

tbursztyka replied:
Ok, I see. That explains why the test passes: it's fully sequential. I have an idea how this could be fixed, as long as the fragments arrive it would not remove the fragment header, just append them to the net_pkt. For each fragment received it would recompute the offset vs size on the whole list, and if all seems to be fine it would rearrange the list according to offset, and get rid of the fragment hdr on each then.

It would be far less costly than the legacy net_pkt_write (which is doomed to be removed)

rveerama1 replied:
> as long as the fragments arrive it would not remove the fragment header, just append them to the net_pkt.

Problem here, one fragment header size is 4 bytes (which should be beginning of the packet) and rest of fragments header size is 5 bytes. It's not always same fragment header size to remove later.

rveerama1 replied:
> Ok, I see. That explains why the test passes: it's fully sequential.

Yep, we don't have random fragment tests.

tbursztyka replied:
That's ok, the info on the type of header is located on the first byte of the buffer, so header removal would take that into account to know the size of hdr to remove.

rveerama1 replied:
Yes

rveerama1 commented on subsys/net/l2/ieee802154/ieee802154_fragment.c:
```diff
@@ -334,34 +345,23 @@ static inline struct frag_cache *get_reass_cache(u16_t size, u16_t tag)
-			 */
-			net_pkt_frag_unref(pkt->frags);
-			pkt->frags = NULL;
+	if (offset != net_pkt_get_len(pkt)) {
```
Look at my previous comment. Sometime offset could be more than what currently received total packet length.

Commit:
net/6lo: Add a way to get the difference of hdr size after uncompression

Since we don't want to redo the complex calculation, let's reuse the
uncompress functions, adding a dry_run possibility to those.

Signed-off-by: Tomasz Bursztyka <tomasz.bursztyka@linux.intel.com>

tbursztyka commented:
Got something. It was a tiny bit more complex than I thought due to the 6lo compressed header. It is not possible to compute the real packet size without decompressing it, but since decompressing for real would be killing the logic I tried to get, I added a 6lo function that makes a dry run of decompression, returning only the header size difference. It is a bit ugly to have this if (!dry_run) everywhere now, but that was the simplest way to achieve what I wanted. Code is also a bit more complex in fragmentation part, but it is in fact simple: it appends all fragments to pkt, the only specific one being the very first fragment (identified as such, that is) which takes the very first place in the list. Once all seems received (size fits the target size), reconstructing is done by reordering the fragments according to there offset. I added an  idea in the commit message how it could be optimized (i.e. running the dry run decompression the least possible).

@rlubos let me know if that works for you.

rlubos commented on subsys/net/ip/6lo.c:
```diff
@@ -1431,3 +1577,24 @@ bool net_6lo_uncompress(struct net_pkt *pkt)
+		return -1;
+	}
+
+	NET_WARN("Nothing to decompress");
```
I'm getting  quite a lot of these in my logs. Is a presence of non-compressed header so unusual that it needs a warning level?

tbursztyka replied:
A forgotten message for debugging, sorry

rlubos reviewed (commented):
Seems to work fine, one question left regarding some warnings.

rlubos reviewed (approved):
Thank you, works very well.

rveerama1 reviewed (approved):
Other than dry_run LGTM, approving now. Later I will take a look if I could avoid this dry run and add tests for random order fragments.
